bin/pg_tapgen

```diff
@@ -251,27 +251,27 @@ sub sequences_are {
 sub functions_are {
     my $schema = shift;
     my $sth = $dbh->prepare(q{
-        SELECT p.proname, md5(p.prosrc) as md5, proargtypes::text
+        SELECT p.proname, md5(p.prosrc) as md5, oidvectortypes(proargtypes) as proargs
           FROM pg_catalog.pg_proc p
           JOIN pg_catalog.pg_namespace n ON p.pronamespace = n.oid
          WHERE n.nspname = ?
          ORDER BY p.proname
     });
     my $allfuncs = $dbh->selectall_arrayref($sth, undef, $schema);
     return unless $allfuncs && @$allfuncs;
-    my @funcs=map{$_->[0]} @$allfuncs;
+    my %funcs=map{$_->[0] => undef} @$allfuncs;
```

Comment:
Why do this and sort below? The query above already has an `ORDER BY` clause.

Reply:
Pushing all the function names into a hash gives us distinct function names. However, Perl hashes are not order-preserving, so we have to sort the results to maintain stable output.


To expand further, say we are generating tests for schema `funcs`, which has the following defined:

```
# \df funcs.
                                                                                               List of functions
 Schema |          Name           |                                                      Result data type                                                      |         Argument data types          |  Type  
--------+-------------------------+----------------------------------------------------------------------------------------------------------------------------+--------------------------------------+--------
 funcs  | num_days                | integer                                                                                                                    | date                                 | normal
 funcs  | num_days                | integer                                                                                                                    | integer, integer                     | normal
```

Currently, this results in the following:
```
SELECT functions_are('funcs', ARRAY[
    'num_days',
    'num_days'
]);
```

Applying this patch instead gives us:
```
SELECT functions_are('funcs', ARRAY[
    'num_days'
]);
```

As `functions_are` does not check that the correct number of overloads of a given function name are present this is functionally identical, more correct and easier to read. There's no good reason to run a separate and largely duplicated query with a `DISTINCT`, so tweaking the code to get the data seems reasonable to me.

Reply:
If you want distinct function names (and I don't understand why you would, since you can have functions with the same names, as they're polymorphic), use the `DISTINCT ON ()` syntax in the query, instead. No need to do any sorting at all in Perl.

Reply:
Ah, I see. In that case, I suggest instead

``` perl
my @funcs = do {
    my %seen;
    grep { !$seen{$_}++ } map { $_->[0] } @$allfuncs;
};
```

Which returns only unique names and preservers the order.

Reply:
That's much cleaner, I pushed up a change that does it that way instead.